File: app/src/main/java/com/philsoft/metrotripper/database/DataProvider.java
package com.philsoft.metrotripper.database;

import android.content.Context;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;

import com.noveogroup.android.log.Logger;
import com.noveogroup.android.log.LoggerManager;
import com.philsoft.metrotripper.database.contracts.StopContract;
import com.philsoft.metrotripper.model.Stop;

import java.util.List;


public class DataProvider {

	private final Logger log = LoggerManager.getLogger(getClass());

	private DatabaseHelper dbHelper;
	private SQLiteDatabase db;


	public DataProvider(Context context) {
		dbHelper = DatabaseHelper.getInstance(context);
		db = dbHelper.getWritableDatabase();
	}

	public Stop getStopById(long stopId) {
		Stop stop = null;
		String where = StopContract.STOP_ID + "=?";
		String[] whereArgs = new String[] { String.valueOf(stopId) };
		Cursor cursor = db.query(StopContract.TABLE_NAME, StopContract.ALL_COLUMNS, where, whereArgs, null, null, null);
		try {
			List<Stop> stops = ModelFactory.buildStops(cursor);
			if (!stops.isEmpty()) {
				stop = stops.get(0);
			}
		} finally {
			cursor.close();
		}
		return stop;
	}

	public List<Stop> getClosestStops(double lat, double lon, int max) {
		List<Stop> stops;

		String sql = "SELECT  * "
				+ " FROM " + StopContract.TABLE_NAME
				+ " ORDER BY ABS(" + StopContract.STOP_LAT + " - " + lat + ") + ABS(" + StopContract.STOP_LON + " - " + lon + ") ASC"
				+ " LIMIT " + max;

		Cursor cursor = db.rawQuery(sql, null);
		try {
			stops = ModelFactory.buildStops(cursor);
		} finally {
			cursor.close();
		}
		return stops;
	}
}

File: app/src/main/java/com/philsoft/metrotripper/database/DataProvider.kt
package com.philsoft.metrotripper.database

import android.content.Context
import com.philsoft.metrotripper.database.contracts.StopContract
import com.philsoft.metrotripper.model.Stop
import org.jetbrains.anko.db.rowParser
import org.jetbrains.anko.db.select


class DataProvider(context: Context) {
    private val dbHelper: DatabaseHelper = DatabaseHelper.getInstance(context)

    companion object {
        val stopParser = rowParser { stopId: Long,
                                     stopName: String,
                                     stopDesc: String,
                                     stopLat: Double,
                                     stopLon: Double,
                                     wheelchairBoarding: Int,
                                     stopUrl: String ->
            Stop(stopId, stopName, stopDesc, stopLat, stopLon, wheelchairBoarding, stopUrl)
        }
    }

    fun getStopById(stopId: Long): Stop? {
        return dbHelper.use {
            val stops = select(StopContract.TABLE_NAME)
                    .whereSimple("${StopContract.STOP_ID} = ?", stopId.toString())
                    .parseList(stopParser)
            return@use if (!stops.isEmpty()) stops[0] else null
        }
    }

    fun getClosestStops(lat: Double, lon: Double, max: Int): List<Stop> {
        return dbHelper.use {
            return@use select(StopContract.TABLE_NAME)
                    .orderBy("ABS(${StopContract.STOP_LAT} - $lat) + ABS(${StopContract.STOP_LON} - $lon)")
                    .limit(max)
                    .parseList(stopParser)
        }
    }
}
